File: csharp/11_Pull/ValuesOfMonthWithCaching.cs
using System;

namespace Pull.Months
{
    class ValuesOfMonthWithCaching : ValuesOfMonth
    {
        private int? balanceLazy = null;
        private int? averageBalanceLazy = null;

        public ValuesOfMonthWithCaching(DateTime dateOfMonth, IValuesOfMonth precedingMonth)
            : base(dateOfMonth, precedingMonth) { }

        public override int Balance
        {
            get
            {
                if (balanceLazy == null)
                {
                    balanceLazy = base.Balance;
                }
                return balanceLazy.Value;
            }
        }

        public override int AverageBalance
        {
            get
            {
                if (averageBalanceLazy == null)
                {
                    averageBalanceLazy = base.AverageBalance;
                }
                return averageBalanceLazy.Value;
            }
        }
    }
}

File: java/10Pull/src/pull/months/ValuesOfMonthWithCaching.java
package pull.months;

import java.util.List;

import org.joda.time.LocalDate;

import common.Transaction_API;

public class ValuesOfMonthWithCaching extends ValuesOfMonth {

	private Integer balanceLazy = null;
	private Integer averageBalanceLazy = null;

	public ValuesOfMonthWithCaching(LocalDate date, IValuesOfMonth month, List<Transaction_API> filteredTransactions) {
		super(date, month, filteredTransactions);
	}

	@Override
	public int getBalance() {
		if (balanceLazy == null) {
			balanceLazy = Integer.valueOf(super.getBalance());
		}
		return balanceLazy.intValue();
	}

	@Override
	public int getAverageBalance() {
		if (averageBalanceLazy == null) {
			averageBalanceLazy = Integer.valueOf(super.getAverageBalance());
		}
		return averageBalanceLazy.intValue();
	}
}
